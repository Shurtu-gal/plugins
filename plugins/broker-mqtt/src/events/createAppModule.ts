import {
  CreateServerAppModuleParams,
  DsgContext,
} from "@amplication/code-gen-types";
import { builders } from "ast-types";
import { join, resolve } from "path";
import { staticsPath } from "../constants";
import { print, readFile } from "@amplication/code-gen-utils";

export const beforeCreateAppModule = async (
  context: DsgContext,
  eventParams: CreateServerAppModuleParams,
) => {
  const { templateMapping, modulesFiles } = eventParams;
  const moduleFilePath = resolve(staticsPath, "mqtt.module.ts");
  const moduleFile = await readFile(moduleFilePath);

  const mqttModuleId = builders.identifier("MqttModule");

  const importArray = builders.arrayExpression([
    mqttModuleId,
    ...eventParams.templateMapping["MODULES"].elements,
  ]);

  templateMapping["MODULES"] = importArray;

  const importFile = {
    code: print(moduleFile).code,
    path: join(
      context.serverDirectories.messageBrokerDirectory,
      "mqtt.module.ts",
    ),
  };

  modulesFiles.set(importFile);

  return eventParams;
};
