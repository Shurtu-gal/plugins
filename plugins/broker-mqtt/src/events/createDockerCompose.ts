import {
  CreateServerDockerComposeDevParams,
  CreateServerDockerComposeParams,
  DsgContext,
} from "@amplication/code-gen-types";
import { getPluginSettings } from "../utils";
import {
  emqxDockerCompose,
  hiveMqDockerCompose,
  hivemqCommunityDockerCompose,
  mosquittoDockerCompose,
} from "../constants";

export const beforeCreateDockerComposeFile = (type: "PROD" | "DEV") => {
  return (
    context: DsgContext,
    eventParams:
      | CreateServerDockerComposeDevParams
      | CreateServerDockerComposeParams,
  ) => {
    const { updateProperties } = eventParams;
    const settings = getPluginSettings(context.pluginInstallations);

    switch (settings.mqttBroker) {
      case "hivemq":
        updateProperties.push(hivemqCommunityDockerCompose);
        break;
      case "mosquitto":
        updateProperties.push(mosquittoDockerCompose);
        break;
      case "hivemq-enterprise":
        updateProperties.push(hiveMqDockerCompose);
        break;
      case "emqx":
        updateProperties.push(emqxDockerCompose);
        break;
    }

    if (type === "PROD") {
      updateProperties.push({
        services: {
          migrate: {
            depends_on: {
              "mqtt-broker": {
                condition: "service_healthy",
              },
            },
          },
        },
      });
    }

    return eventParams;
  };
};
