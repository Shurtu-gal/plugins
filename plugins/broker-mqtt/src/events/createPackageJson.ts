import {
  CreateServerPackageJsonParams,
  DsgContext,
} from "@amplication/code-gen-types";
import { serverPackageJsonValues } from "../constants";

export const beforeCreateServerPackageJson = (
  context: DsgContext,
  eventParams: CreateServerPackageJsonParams,
): CreateServerPackageJsonParams => {
  eventParams.updateProperties.push(serverPackageJsonValues);

  return eventParams;
};
