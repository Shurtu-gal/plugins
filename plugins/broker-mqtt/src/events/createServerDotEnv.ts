import {
  CreateServerDotEnvParams,
  DsgContext,
} from "@amplication/code-gen-types";
import { getPluginSettings } from "../utils";

export const beforeCreateServerDotEnv = (
  context: DsgContext,
  eventParams: CreateServerDotEnvParams,
): CreateServerDotEnvParams => {
  const settings = getPluginSettings(context.pluginInstallations);
  const { resourceInfo } = context;
  const {
    mqttPort,
    mqttPassword,
    mqttUsername,
    mqttClientId,
    mqttBrokerHost,
    mqttWsPort,
    mqttWebUiPort,
  } = settings;
  const { envVariables } = eventParams;

  const clientId = mqttClientId || `broker-mqtt-${resourceInfo?.id}`;

  envVariables.push({
    MQTT_BROKER_HOST: mqttBrokerHost || "localhost",
    MQTT_WEB_UI_PORT: mqttWebUiPort?.toString() || "8080",
    MQTT_PORT: mqttPort?.toString() || "1883",
    MQTT_WS_PORT: mqttWsPort?.toString() || "8073",
    MQTT_USERNAME: mqttUsername,
    MQTT_PASSWORD: mqttPassword,
    MQTT_CLIENT_ID: mqttClientId || clientId,
  });

  return eventParams;
};
