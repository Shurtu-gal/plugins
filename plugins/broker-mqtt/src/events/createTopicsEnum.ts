import {
  CreateMessageBrokerTopicsEnumParams,
  DsgContext,
  EnumMessagePatternConnectionOptions,
  ModuleMap,
  ServiceTopics,
} from "@amplication/code-gen-types";
import { join } from "path";
import { templatesPath } from "../constants";
import { getClassDeclarationById } from "../util/ast";
import { builders, namedTypes } from "ast-types";
import { print, readFile } from "@amplication/code-gen-utils";
import { pascalCase } from "pascal-case";

export const afterCreateTopicsEnum = async (
  context: DsgContext,
  eventParams: CreateMessageBrokerTopicsEnumParams,
  modules: ModuleMap,
): Promise<ModuleMap> => {
  const templatePath = join(templatesPath, "mqtt.controller.template.ts");
  const template = await readFile(templatePath);
  const controllerId = builders.identifier("MqttController");
  const controllerDeclaration = getClassDeclarationById(template, controllerId);
  const { serverDirectories, serviceTopics } = context;

  await addEventPattern(controllerDeclaration, serviceTopics);

  const controllerPath = join(
    serverDirectories.srcDirectory,
    "mqtt",
    "mqtt.controller.ts",
  );

  const controllerModule = {
    code: print(template).code,
    path: controllerPath,
  };

  modules.set(controllerModule);

  return modules;
};

export const addEventPattern = async (
  controllerDeclaration: namedTypes.ClassDeclaration,
  serviceTopics?: ServiceTopics[],
) => {
  serviceTopics?.forEach((serviceTopic) => {
    serviceTopic.patterns.forEach((pattern) => {
      if (!pattern.topicName) {
        throw new Error(`Topic name not found for topic id ${pattern.topicId}`);
      }

      // Only need to add an event pattern for recieve topics
      if (pattern.type !== EnumMessagePatternConnectionOptions.Receive) return;

      // Step 1: Make decorators (Eventpattern, Ctx, Payload)
      const patternDecorator = builders.decorator(
        builders.callExpression(builders.identifier("EventPattern"), [
          builders.stringLiteral(pattern.topicName),
        ]),
      );

      const payloadDecorator = builders.decorator(
        builders.callExpression(builders.identifier("Payload"), []),
      );

      const contextDecorator = builders.decorator(
        builders.callExpression(builders.identifier("Ctx"), []),
      );

      // Step 2: Make parameters to class function
      const value = builders.identifier.from({
        name: "value",
        typeAnnotation: builders.tsTypeAnnotation(
          builders.tsTypeReference(
            builders.identifier("string | Record<string, any> | null"),
          ),
        ),
      });

      const context = builders.identifier.from({
        name: "context",
        typeAnnotation: builders.tsTypeAnnotation(
          builders.tsTypeReference(builders.identifier("MqttContext")),
        ),
      });

      //@ts-expect-error - decorators is defined in the type
      value.decorators = [payloadDecorator];
      //@ts-expect-error - decorators is defined in the type
      context.decorators = [contextDecorator];

      // Step 3: Make the function
      const listener = builders.classMethod.from({
        async: true,
        key: builders.identifier(`on${pascalCase(pattern.topicName)}`),
        params: [value, context],
        decorators: [patternDecorator],
        returnType: builders.tsTypeAnnotation(
          builders.tsTypeReference(
            builders.identifier("Promise"),
            builders.tsTypeParameterInstantiation([builders.tsVoidKeyword()]),
          ),
        ),
        body: builders.blockStatement([
          builders.variableDeclaration("const", [
            builders.variableDeclarator(
              builders.identifier("originalPacket"),
              builders.callExpression(
                builders.memberExpression(
                  builders.identifier("context"),
                  builders.identifier("getPacket"),
                ),
                [],
              ),
            ),
          ]),
          builders.variableDeclaration("const", [
            builders.variableDeclarator(
              builders.identifier("topic"),
              builders.callExpression(
                builders.memberExpression(
                  builders.identifier("context"),
                  builders.identifier("getTopic"),
                ),
                [],
              ),
            ),
          ]),
        ]),
      });

      controllerDeclaration.body.body.push(listener);
    });
  });
};
