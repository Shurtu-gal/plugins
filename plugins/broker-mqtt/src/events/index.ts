export * from "./createServerDotEnv";
export * from "./createPackageJson";
export * from "./createAppModule";
export * from "./createDockerCompose";
