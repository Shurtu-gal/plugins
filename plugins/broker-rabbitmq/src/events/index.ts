export * from "./createServer";
export * from "./createMessageBrokerService";
export * from "./createConnectMicroservices";
