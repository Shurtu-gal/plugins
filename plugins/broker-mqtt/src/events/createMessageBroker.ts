import { DsgContext, ModuleMap } from "@amplication/code-gen-types";
import { join, resolve } from "path";
import { staticsPath, templatesPath } from "../constants";
import { print, readFile } from "@amplication/code-gen-utils";
import { EnumResourceType } from "@amplication/code-gen-types/src/models";
import { pascalCase } from "pascal-case";
import { interpolate } from "../util/ast";
import { builders } from "ast-types";

export const afterCreateMessageBrokerClientOptionsFactory = async (
  context: DsgContext,
): Promise<ModuleMap> => {
  const { serverDirectories } = context;
  const filePath = resolve(staticsPath, "generateClientOptions.ts");
  const file = await readFile(filePath);
  const generateFileName = "generateClientOptions.ts";

  const path = join(serverDirectories.messageBrokerDirectory, generateFileName);
  const modules = new ModuleMap(context.logger);
  await modules.set({ code: print(file).code, path });
  return modules;
};

export const afterCreateMessageBrokerNestJSModule = async (
  context: DsgContext,
): Promise<ModuleMap> => {
  const filePath = resolve(staticsPath, "mqtt.module.ts");
  const file = await readFile(filePath);
  const generateFileName = "mqtt.module.ts";

  const moduleFile = {
    code: print(file).code,
    path: join(
      context.serverDirectories.messageBrokerDirectory,
      generateFileName,
    ),
  };

  const modules = new ModuleMap(context.logger);
  await modules.set(moduleFile);
  return modules;
};

export const afterCreateMessageBrokerService = async (
  context: DsgContext,
): Promise<ModuleMap> => {
  const { otherResources, logger } = context;

  const servicePath = join(
    context.serverDirectories.messageBrokerDirectory,
    "mqtt.producer.service.ts",
  );

  let messageBrokerName =
    otherResources?.find(
      (resource) => resource.resourceType === EnumResourceType.MessageBroker,
    )?.resourceInfo?.name ?? null;

  if (!messageBrokerName) {
    logger.warn(
      "Message broker name not found. Did you forget to add a message broker resource? Adding a default name mqtt",
    );
    messageBrokerName = "mqtt";
  }

  const templatePath = join(templatesPath, "mqtt.producer.service.template.ts");
  const template = await readFile(templatePath);
  const templateMapping = {
    BROKER_TOPICS: builders.identifier(
      pascalCase(messageBrokerName) + "Topics",
    ),
  };

  interpolate(template, templateMapping);

  const modules = new ModuleMap(logger);
  await modules.set({ code: print(template).code, path: servicePath });

  const typesPath = join(
    context.serverDirectories.messageBrokerDirectory,
    "types.ts",
  );

  const typesFile = await readFile(join(staticsPath, "types.ts"));
  await modules.set({ code: print(typesFile).code, path: typesPath });

  return modules;
};
