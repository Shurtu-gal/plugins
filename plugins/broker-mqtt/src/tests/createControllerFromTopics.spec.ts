import {
  BuildLogger,
  CreateMessageBrokerTopicsEnumParams,
  DsgContext,
  EnumMessagePatternConnectionOptions,
  ModuleMap,
} from "@amplication/code-gen-types";
import { mock } from "jest-mock-extended";
import { afterCreateTopicsEnum } from "../events";

describe("Testing the creation of topic subscription in controller", () => {
  let context: DsgContext;
  let logger: BuildLogger;
  let modules: ModuleMap;
  let eventParams: CreateMessageBrokerTopicsEnumParams;

  beforeEach(() => {
    eventParams = mock<CreateMessageBrokerTopicsEnumParams>();
    logger = mock<BuildLogger>();
    context = mock<DsgContext>({
      pluginInstallations: [
        {
          npm: "@amplication/plugin-broker-mqtt",
        },
      ],
      serverDirectories: {
        srcDirectory: "src",
        messageBrokerDirectory: "default-message-broker",
      },
      serviceTopics: [],
      logger: {
        info: jest.fn(),
        error: jest.fn(),
        warn: jest.fn(),
      },
      otherResources: [],
    });
    modules = new ModuleMap(logger);
  });

  it("should create a controller without topics if no resources are found", async () => {
    await afterCreateTopicsEnum(context, eventParams, modules);
    expect(modules.modules()).toMatchSnapshot();
  });

  it("should create a controller with topics if resources are found", async () => {
    context.serviceTopics = [
      {
        enabled: true,
        id: "test-topic-id",
        messageBrokerId: "test-broker-id",
        patterns: [
          {
            type: EnumMessagePatternConnectionOptions.Receive,
            topicId: "test-topic-id",
            topicName: "test-topic-name",
          },
          {
            type: EnumMessagePatternConnectionOptions.Send,
            topicId: "should-not-be-added",
            topicName: "should-not-be-added",
          },
          {
            type: EnumMessagePatternConnectionOptions.Receive,
            topicId: "should-be-added",
            topicName: "should-be-added",
          },
        ],
      },
    ];

    await afterCreateTopicsEnum(context, eventParams, modules);
    expect(modules.modules()).toMatchSnapshot();
  });

  it("should throw an error if no topic name is found", async () => {
    context.serviceTopics = [
      {
        enabled: true,
        id: "test-topic-id",
        messageBrokerId: "test-broker-id",
        patterns: [
          {
            type: EnumMessagePatternConnectionOptions.Receive,
            topicId: "test-topic-id",
          },
        ],
      },
    ];

    await expect(
      afterCreateTopicsEnum(context, eventParams, modules),
    ).rejects.toThrowErrorMatchingSnapshot();
  });
});
