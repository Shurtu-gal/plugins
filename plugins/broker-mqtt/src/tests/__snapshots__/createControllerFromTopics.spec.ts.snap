// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing the creation of topic subscription in controller should create a controller with topics if resources are found 1`] = `
[
  {
    "code": "import { Controller } from "@nestjs/common";
import { Ctx, EventPattern, MqttContext, Payload } from "@nestjs/microservices";

@Controller("mqtt-controller")
export class MqttController {
  @EventPattern(test-topic-name)
  async onTestTopicName(
    @Payload()
    value: string | Record<string, any> | null,
    context: MqttContext
  ): Promise<void> {
    const message = context.getPacket();
    const topic = context.getTopic();
  }

  @EventPattern(should-be-added)
  async onShouldBeAdded(
    @Payload()
    value: string | Record<string, any> | null,
    context: MqttContext
  ): Promise<void> {
    const message = context.getPacket();
    const topic = context.getTopic();
  }
}
",
    "path": "src/mqtt/mqtt.controller.ts",
  },
]
`;

exports[`Testing the creation of topic subscription in controller should create a controller without topics if no resources are found 1`] = `
[
  {
    "code": "import { Controller } from "@nestjs/common";
import { Ctx, EventPattern, MqttContext, Payload } from "@nestjs/microservices";

@Controller("mqtt-controller")
export class MqttController {}
",
    "path": "src/mqtt/mqtt.controller.ts",
  },
]
`;

exports[`Testing the creation of topic subscription in controller should throw an error if no topic name is found 1`] = `"Topic name not found for topic id test-topic-id"`;
